Add support for key logging callbacks.

Lukasa opened the pull request:
Follows on from #570.

This PR adds support for registering callbacks on an SSL_CTX object that are called whenever keying information is negotiated. These callbacks are called with the SSL object and also a string which is formatted in the style of NSS's SSLKEYLOGFILE entries. This callback has substantial value for debugging applications that are using OpenSSL to provide their transport, as it allows programs like Wireshark to decrypt the TLS traffic and inspect the content.

I've tested this patch live and it does appear to work. However, I'd consider it incomplete at the moment because it currently has no tests. I'd like a bit of insight from the OpenSSL developers: what form of testing would you like for this functionality? I'm happy to write whatever tests are required, but it'd be good to know what you're expecting so I don't spend hours writing tests that are ultimately not what you're looking for.

I should note as well, for the sake of openness, that this is not a clean-room implementation. Many of the functions are directly inspired by the work @zakjan did in #570, as are their placement in the codebase. Additionally, substantial inspiration was drawn from BoringSSL, which defined this API and whose implementation I kept an eye on. In fact, I should note that BoringSSL has a somewhat more general implementation with support for TLS 1.3 which I did not bring over due to OpenSSL's current lack of 1.3 support. For this reason, I think we should consult @davidben and his colleagues to ensure that we're in a safe place, copyright-wise, with this work.

zakjan commented:
I didn't have time to continue with proper implementation of my patch, thanks for reopening it. It is an important feature to have for debugging. Can you reply to @richsalz 's comments from #570?

Lukasa commented:
@zakjan So I believe this does address @richsalz's concerns. Specifically, this patch no longer checks for environment variables. Instead, the application is responsible for hooking up the debug output however it sees fit. For example, applications could use config files instead of environment variables. And any application that wants to emulate SSLKEYLOGFILE can do so, by listening for that environment variable itself.

richsalz commented:
So the first thing that has to happen is we'll need signed CLA's from @Lukasa and @zakjan.
Then we need to know exactly which files from boringSSL were used, so we can get Google to approve that (they will, we just need specifics).
As for a test, adding something to sslapitest is the first place to start.

No hurry, since it's awhile before our next release.

Lukasa commented:
@richsalz I've already sent in a signed CLA: I sent it to legal@opensslfoundation.org on the 5th of June, 2016.

As to which files I consulted in BoringSSL, I looked at the following files:

- `include/openssl/ssl.h`
- `ssl/handshake_client.c`
- `ssl/internal.h`
- `ssl/s3_both.c`
- `ssl/ssl_lib.c`
- `ssl/tls13_enc.c` (I didn't use anything from here, but wanted to confirm that this was TLS 1.3 only and didn't need cover here yet)

I'll take a look at the testing now, and will follow up if I have any other queries. Thanks @richsalz!

richsalz commented:
Thanks for the info; yes we have your CLA.

agl commented:
Please consider any work from BoringSSL included here to be covered by our existing CLA with OpenSSL.

Lukasa commented:
Ok, I've pushed up an initial test. This test roughly validates that the output is sensible. Right now it doesn't check that it's *right*, though, so on that front I have a few questions:

- I see that we have `SSL_SESSION_get_master_key`, which I can use to validate that the master key is correct. Is that worth my doing?
- We also have `SSL_get_cleint_random`, which I can use. Again, is that worthwhile?
- Is it worth parametrizing this across a few other cipher suites?
- I can't find any way to get the pre-master secret without reaching into structures. Is it worth validating that output?

Lukasa commented:
I've actually just gone ahead and improved the testing for the master secret logging. I'm still not sure how we can test the output for the RSA CKE, so suggestions there are welcome.

richsalz commented:
Thanks, @agl, just needed it explicit :)  So depending on whether or not a CLA is needed from @zakjan.

zakjan commented:
I have sent my signed CLA today. However I can't say #570 is my own work, because it is just the patch taken from https://rt.openssl.org/Ticket/Display.html?id=4201#txn-59377, and rebased to the latest master at that time. I hope it should be covered by BoringSSL CLA?

Lukasa commented:
Ok, I have green ticks now. This is probably ready for actual code review whenever an OpenSSL dev has time to do it.

davidben commented:
The safer 'n' variants of `strcpy` and `sprintf` would probably be better to use. (Except I believe OpenSSL uses `BIO_snprintf` instead of `snprintf`.)

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
 #endif
+
+void SSL_CTX_set_keylog_callback(SSL_CTX *ctx,
+                                 void (*cb)(const SSL *ssl,
```
please create a typedef for this function pointer.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+  ctx->keylog_callback = cb;
+}
+
+void (*SSL_CTX_get_keylog_callback(const SSL_CTX *ctx))(const SSL *ssl,
```
and use the typedef here, too :)

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+  size_t out_len = 0;
+  size_t i;
+
+  if (ssl->ctx->keylog_callback == NULL) {
```
remove the extra curly braces

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+   * - A trailing NULL.
+   */
+  out_len = 4 + 16 + 1 + premaster_len*2 + 1;
+  if (!(out = cursor = OPENSSL_malloc(out_len))) {
```
test against NULL not using !

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+    sprintf(cursor, "%02x", encrypted_premaster[i]);
+    cursor += 2;
+  }
+  *cursor = ' ';
```
merge, using post-increment.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+  size_t out_len = 0;
+  size_t i;
+
+  if (ssl->ctx->keylog_callback == NULL) {
```
curly braces.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -4271,3 +4271,125 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+      sprintf(cursor, "%02x", client_random[i]);
+      cursor += 2;
+  }
+  *cursor = ' ';
```
post-increment.

richsalz commented on ssl/ssl_locl.h:
```diff
@@ -2080,6 +2086,25 @@ __owur const EVP_MD *ssl_md(int idx);
 __owur const EVP_MD *ssl_handshake_md(SSL *s);
 __owur const EVP_MD *ssl_prf_md(SSL *s);
 
+/* ssl_log_rsa_client_key_exchange logs |premaster| to the SSL_CTX associated
```
open-comment on a line by itself.

richsalz commented on ssl/statem/statem_clnt.c:
```diff
@@ -2206,6 +2206,11 @@ static int tls_construct_cke_rsa(SSL *s, WPACKET *pkt, int *al)
 # endif
 
+    /* Log the premaster secret, if logging is enabled. */
+    if (!ssl_log_rsa_client_key_exchange(s, encdata, enclen, pms, pmslen)) {
```
curly braces.

richsalz commented on ssl/statem/statem_lib.c:
```diff
@@ -100,6 +100,13 @@ int tls_construct_finished(SSL *s, WPACKET *pkt)
+    /* Log the master secret, if logging is enabled. */
+    if (!ssl_log_master_secret(s, s->s3->client_random, SSL3_RANDOM_SIZE,
+                               s->session->master_key,
+                               s->session->master_key_length)) {
```
curly braces.

Lukasa commented:
I have to restart the build here.

Lukasa commented:
Ok @richsalz, this should be ready for further review. I believe I've addressed all your comments.

richsalz commented:
yes, this is good. +1 from me.  needs another team reviewer.

Lukasa commented:
Updated to resolve merge conflicts.

levitte commented on doc/ssl/SSL_CTX_set_keylog_callback.pod:
```diff
@@ -0,0 +1,46 @@
+
+void(*ssl_ctx_keylog_callback)(const SSL *ssl, const char *line);
+void SSL_CTX_set_keylog_callback(SSL_CTX *ctx, ssl_ctx_keylog_callback cb);
+ssl_ctx_keylog_callback SSL_CTX_get_keylog_callback(const SSL_CTX *ctx);
```
These three lines should be indented with one space

levitte commented on doc/ssl/SSL_CTX_set_keylog_callback.pod:
```diff
@@ -0,0 +1,46 @@
+
+=head1 COPYRIGHT
+
+Copyright 2002-2016 The OpenSSL Project Authors. All Rights Reserved.
```
`s/2002-2016/2016-2017/` at this point.

levitte commented on ssl/ssl_lib.c:
```diff
@@ -4375,3 +4375,118 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+  ssl->ctx->keylog_callback(ssl, (const char *)out);
+  OPENSSL_free(out);
+  return 1;
+}
```
This change needs a bit of a formatting tweak, since we do 4 space indentation.

Lukasa commented:
@levitte I've rebased on top of the current master branch, updated the ordinals, and applied patches to address your feedback.

As to your question about logging key exchanges, as far as I am aware [the NSS key log file format only logs RSA key exchanges](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format). In principle the CLIENT_RANDOM format is used for the others. However, I've definitely seen BoringSSL extend this format a bit: I have no idea if that's part of the formal specification for the key log format. I think it isn't: the custom BoringSSL stuff applies to TLSv1.3, and I'm going to admit enough uncertainty about what's going on in their TLSv1.3 file to suggest that if you want to pursue logging that kind of information that we should tag a BoringSSL developer into this conversation to enlighten us. I'm definitely happy to do the work required to make that function for OpenSSL if we want it: I just don't know enough about what's going on to get started myself.

richsalz commented:
@levitte, once you're happy this gets a +1 from me.

davidben commented:
> In principle the CLIENT_RANDOM format is used for the others. However, I've definitely seen BoringSSL extend this format a bit

No, CLIENT_RANDOM only logs the TLS 1.2 (and below) master secret. This does not make sense with TLS 1.3's key schedule, so the format needed to be extended:
https://bugzilla.mozilla.org/show_bug.cgi?id=1287711
(There are actually a few other line formats in the Wireshark source, but I don't know who emits them.)

But the TLS 1.3 stuff is all moot because nothing consumes it anyway. I don't think anyone's gotten around to implementing the Wireshark half. We implemented it just so there'll be something to work with when someone gets to fleshing out Wireshark's TLS 1.3 dissector.

Lukasa commented:
Thanks @davidben! So my instinct, then, would be to say that this covers enough ground as-is. It'll need revisiting when OpenSSL's TLSv1.3 work is better baked, at which time we'll hopefully have a better idea of what Wireshark needs from the format.

levitte commented on ssl/ssl_lib.c:
```diff
@@ -4344,3 +4344,118 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+    ssl->ctx->keylog_callback(ssl, (const char *)out);
+    OPENSSL_free(out);
+    return 1;
+}
```
These two functions have a *lot* in common.  Would it be possible to do some code reuse rather than code copying, i.e. have a common function that takes some extra parameters, say:

``` C
static int nss_keylog_int(const char *prefix,
                          SSL *ssl,
                          const uint8_t *p1, size_t p1_len,
                          const uint8_t *p2, size_t p2_len)
```

And then have call it like this?

``` C
int ssl_log_rsa_client_key_exchange(SSL *ssl,
                                    const uint8_t *encrypted_premaster,
                                    size_t encrypted_premaster_len,
                                    const uint8_t *premaster,
                                    size_t premaster_len)
{
    if (encrypted_premaster_len < 8) {
        SSLerr(SSL_F_SSL_LOG_RSA_CLIENT_KEY_EXCHANGE, ERR_R_INTERNAL_ERROR);
        return 0;
    }
    return nss_keylog_int("RSA",
                          ssl,
                          encrypted_premaster, 8,
                          premaster, premaster_len);
}

int ssl_log_master_secret(SSL *ssl,
                          const uint8_t *client_random,
                          size_t client_random_len,
                          const uint8_t *master,
                          size_t master_len)
{
    if (client_random_len != 32) {
        SSLerr(SSL_F_SSL_LOG_MASTER_SECRET, ERR_R_INTERNAL_ERROR);
        return 0;
    }
    return nss_keylog_int("CLIENT_RANDOM",
                          ssl,
                          client_random, 32,
                          master, master_len);
}

```

levitte replied:
Come to think of it, when `nss_keylog_int` returns 0, an error `ERR_R_MALLOC_FAILURE` should be generated...

Lukasa replied:
Yup, I can do that. :smile:

levitte commented:
> > In principle the CLIENT_RANDOM format is used for the others. However, I've definitely seen BoringSSL extend this format a bit
>
> No, CLIENT_RANDOM only logs the TLS 1.2 (and below) master secret. 

Note that [the docs](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format) that @Lukasa pointed at says this:

> *The CLIENT_RANDOM format allows other key-agreement algorithms to be logged but is only supported starting with Wireshark 1.8.0.*

It makes no sense to me, though, as the same page specifies very precise sizes:

> `CLIENT_RANDOM <space> <64 bytes of hex encoded client_random> <space> <96 bytes of hex encoded master secret>`

davidben commented:
Right, the size is fixed because it's the master secret. That page was presumably written before TLS 1.3 so it talks about generalizing to key exchanges other than static RSA.

It won't work for TLS 1.3 because there isn't a single secret to log. You can get away with two, but the most natural thing is to use the five traffic secrets.

Lukasa commented on doc/ssl/SSL_CTX_set_keylog_callback.pod:
```diff
@@ -0,0 +1,46 @@
+
+=head1 NAME
+
+SSL_CTX_set_keylog_callback, SSL_CTX_get_keylog_callback, ssl_ctx_keylog_callback - logging TLS key material
```
Does that apply to the entire pod file?

Lukasa replied:
Uh...nevermind, I notice now that this question makes no sense.

levitte commented:
The code looks good to me.  I'm going to wait for the build & test results to approve.

Lukasa commented:
So I'm not entirely able to work out what's going on in these test failures. One thing I think I've identified is that `make doc-nits` seems to have trouble parsing the syntax for function pointers in the pod file, so if you can let me know how you'd like that changed I can do that. The rest of the output is really unclear: there seem to be some tests failing, but they don't appear to be related to this logging functionality as far as I can tell.

levitte commented on doc/ssl/SSL_CTX_set_keylog_callback.pod:
```diff
@@ -0,0 +1,47 @@
+
+ #include <openssl/ssl.h>
+
+ void(*ssl_ctx_keylog_callback)(const SSL *ssl, const char *line);
```
This is meant to be a typedef, methinks.  So:

``` text
 typedef void (*ssl_ctx_keylog_callback)(const SSL *ssl, const char *line);
```

That should fix the doc-nits issue

Lukasa replied:
Yup, that looks right.

levitte commented:
> The rest of the output is really unclear: there seem to be some tests failing, but they don't appear to be related to this logging functionality as far as I can tell.

If you choose the raw log and search for "not ok", you will find this, which does seem related...

``` text
../../_srcdist/test/recipes/90-test_sslapi.t ........... 
1..1
./sslapitest: 121 test cases
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_connect() failed -1, 1
No progress made
Unable to create SSL connection
** test_keylog failed **
--------
./sslapitest: 1 test failed (out of 121)
not ok 1 - running sslapitest
../util/shlib_wrap.sh ./sslapitest ../../_srcdist/apps/server.pem ../../_srcdist/apps/server.pem => 1

#   Failed test 'running sslapitest'
#   at ../../_srcdist/test/recipes/90-test_sslapi.t line 20.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
```

Lukasa commented:
Well spotted @levitte, and this is why we have tests. Looks like I had inverted a conditional when I did the refactor. :grin:

levitte commented:
Cool, good find.  Let's hope the builds go through this time then

richsalz commented:
are the doc-nits only in pages you modified?  if so please post the output here.  if not, ignore the doc-nits failures.

Lukasa commented:
They were, yes, though I've since fixed them with @levitte's help. The output was:

```
(cd .; /usr/bin/perl util/find-doc-nits.pl -n ) >doc-nits
if [ -s doc-nits ] ; then cat doc-nits; exit 1; fi
doc/ssl/SSL_CTX_set_keylog_callback.pod:1: void missing from NAME section
doc/ssl/SSL_CTX_set_keylog_callback.pod:1: ssl_ctx_keylog_callback missing from SYNOPSIS
make: *** [doc-nits] Error 1
```

FdaSilvaYY commented on include/openssl/ssl.h:
```diff
@@ -775,6 +775,25 @@ __owur int SSL_extension_supported(unsigned int ext_type);
+ * A callback for logging out TLS key material. This callback should log out
+ * |line| followed by a newline.
+ */
+typedef void(*ssl_ctx_keylog_callback)(const SSL *ssl, const char *line);
```
Same fix as for POD file , no ?

Lukasa replied:
I think that isn't needed: the typedef is present here.

FdaSilvaYY replied:
I was saying : "Add a space between void and the parenthesis" ... like you do fix the pod file :
Syntax is ok, but it is more readable with it.

Lukasa replied:
Oh, I see. ;) I did that while I was fixing up the case.

FdaSilvaYY commented on test/sslapitest.c:
```diff
@@ -34,6 +41,250 @@ static X509 *ocspcert = NULL;
+    }
+
+    /*
+     * Now do a handshake and check that the logs have been written to.
```
One line comment ? : 
/* comment .. */

Lukasa replied:
That I could certainly do!

richsalz commented:
The typedef should be ```SSL_CTX_...```  uppercase prefix.

levitte commented:
@Lukasa, I'm noticing that this is a looot of commits for a fairly small change.  At some point, it would be good if they were squashed into fewer ones before we merge this change...  say one for the implementation, one for the test, one for the docs?

Lukasa commented:
@levitte Sure, that's not a problem.

A side note, but it looks like the TLSv1.3 tests are failing with this change, and it looks like they're failing for an interesting reason. I'm going to dig into that sometime this morning, so once that's done I'll also rebase down.

levitte commented:
> A side note, but it looks like the TLSv1.3 tests are failing with this change, and it looks like they're failing for an interesting reason. I'm going to dig into that sometime this morning, so once that's done I'll also rebase down.

The master key is computed differently with TLSv1.3.  @mattcaswell or @davidben may be able to fill you in...

Lukasa commented:
Yeah, so @davidben has pointed out already that this logging functionality works differently for TLSv1.3, and most importantly isn't standardised yet. I'd be very nervous about dropping a patch that includes logging that may require updating in the future, mostly because I'm reluctant to add too much maintenance burden to the OpenSSL team if I can avoid it.

I think the sensible way forward for this patch, then, is for me to see if I can detect in which situations I cannot meaningfully log out a master key and to choose not to. That will also require another test to validate that function, but that's not a problem.

Lukasa commented:
@levitte So my proposal is to add a check in `ssl_log_master_secret` that returns early if `ssl->version >= TLS1_3_VERSION`. I'll then add a new test code that validates that we don't log anything in the handshake, and clamp the current test to TLSv1.2 and lower. How does that sound to you?

levitte commented:
That's fine, I guess...  me, I was rather thinking that the `ssl_log_master_secret` call could be moved to the code blocks where `s->session->master_key_length` is assigned a value.  There are two of them, one in `ssl/statem/statem_clnt.c` and one in `ssl/statem/statem_srvr.c`

levitte commented:
... sorry, three, there's `ssl/ssl_asn1.c` as well...

Lukasa commented:
@levitte Is it better to do that, or to add a check on the master key length in the logging function?

levitte commented:
Good question.  Frankly, it's beyond my knowledge to answer to that at this point

Lukasa commented:
Fair enough, I'll wait to see if someone else has an opinion. I have now got an explicitly failing test for the TLSv1.3 configuration, so hopefully we should be able to work this out soonish.

mattcaswell commented:
Not really been following this PR, but I'd make a couple of comments:

> So my proposal is to add a check in ssl_log_master_secret that returns early if ssl->version >= TLS1_3_VERSION

Use the macro `SSL_IS_TLS13()` rather than directly testing `ssl->version`.

I'd also like to point out that in #2259, there are some changes to the storage of the master secret for TLSv1.3. Specifically the master secret is stored in the SSL object (along with the early secret and the handshake secret). The `master_key` field in the SSL_SESSION is used to store the TLSv1.3 resumption master secret only - and only gets filled in when a session is actually established (which may be sometime after the handshake is completed). This may not have any impact for you if you are going to disable your code from working in TLSv1.3 anyway - but perhaps it will have an influence on your design.

Lukasa commented:
Thanks @mattcaswell.

So, for TLSv1.3 BoringSSL logs out four secrets that it calls "CLIENT_HANDSHAKE_TRAFFIC_SECRET", "SERVER_HANDSHAKE_TRAFFIC_SECRET", "CLIENT_TRAFFIC_SECRET_0", and "SERVER_TRAFFIC_SECRET_0". These are logged out instead of the CLIENT_RANDOM log that this PR currently logs, and they are logged at the time they are derived.

In BoringSSL the CLIENT_RANDOM log is fired at the same time as in this PR: namely, when the finished message is being generated. The reason Boring doesn't have the same problem we're having here is that they have a separate "finished" function for TLSv1.3 that does not include this logging. As a result, they don't need a "don't do this for TLSv1.3" check, because that check is implicit in the use of an entirely different function.

Now, I'm aware that TLSv1.3 work is ongoing in OpenSSL, which means that the architecture is going to change at some point in the future. So it's entirely possible that this problem will go away, and that we could add a quick `if (!SSL_IS_TLS13(ssl))` to the `ssl_log_master_secret` function that would eventually become superfluous.

On the other hand, if the folks working on OpenSSL's TLSv1.3 support believe that they too will write a new `tls13_finished` function that does not have the same code path, we can just wait for that change to happen before updating this PR and the problem will go away. Another advantage of this "wait and see" approach is that we can then use our newly-genericized function to log out the same 4 secrets as BoringSSL does, giving us the same logging capability as BoringSSL right out of the gate.

Does anyone have a particular viewpoint on which way we should go here?

mattcaswell commented:
> On the other hand, if the folks working on OpenSSL's TLSv1.3 support believe that they too will write a new tls13_finished function that does not have the same code path, we can just wait for that change to happen before updating this PR and the problem will go away.

Well I've already implemented all the "Finished" message changes needed for TLSv1.3, so I'm not currently expecting to need a separate function for this. I can't make any guarantees though.

Lukasa commented:
@mattcaswell In that case, I think the question becomes whether it's ever sensible to log out the master secret for TLSv1.3. @davidben suggested the answer was no. Assuming that's true, we probably want to not allow logging the master secret for TLSv1.3.

mattcaswell commented:
> Assuming that's true, we probably want to not allow logging the master secret for TLSv1.3.

I think that's reasonable for now.

Lukasa commented:
Ok, I'll put a quick return in the master secret log. I'll also add a comment there that explains why we're skipping it, because this may come up again for TLSv1.4 (if and when that happens), and we should make it easy for a future programmer to quickly decide whether they need to also exclude TLSv1.4 (and higher).

richsalz commented:
Yeah, don't log the 1.3 master secret.    Let's wait and see what wireshark needs/want.

davidben commented:
(To clarify, the TLS 1.3 scheme is to log five secrets, not four. There will also be `CLIENT_EARLY_TRAFFIC_SECRET` for 0-RTT. They're just the five traffic secrets in the [key schedule](https://tlswg.github.io/tls13-spec/#rfc.section.7.1) because 1.3 does a lot more key changes and has split client/server ladders.)

Lukasa commented:
Ok, looks like the tests pass. Per @levitte, I'm going to squash these commits down now.

Lukasa commented:
Ok, that's done. This should be ready to go @levitte.

levitte commented on ssl/statem/statem_clnt.c:
```diff
@@ -2513,6 +2513,10 @@ static int tls_construct_cke_rsa(SSL *s, WPACKET *pkt, int *al)
+    /* Log the premaster secret, if logging is enabled. */
+    if (!ssl_log_rsa_client_key_exchange(s, encdata, enclen, pms, pmslen))
+        goto err;
+
```
I'm thinking this should move down below the next condition, so it ends up just before this:

``` C
     s->s3->tmp.pms = pms;
```

That's to make sure it only gets logged if *everything* else worked out fine.

Lukasa replied:
Done. =)

richsalz commented:
aargh, sorry.  ```SSL_CTX_set_keylog_callback``` needs a newline before the curly.
And can you change the typedef to end with ```_cb``` for consistency?
And in the manpage put the typedef first or last, and put a blank line separating it from the functions.
Thanks for your forebearance!

Lukasa commented:
Ack, sorry, somehow I missed the email for this comment. Yup, I'll do all of that now.

Lukasa commented:
Ok @richsalz, I've done that. Rather than end the typedef in `_cb` I've ended it in `_cb_func`, which seems to be more consistent with the ALPN and NPN callback typedefs. Let me know if that wasn't what you wanted.

Lekensteyn commented:
@davidben Some of the keylog formats for Wireshark are documented in this [post][1]. Support for TLS 1.3 including decryption is tracked at https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=12779 (initial focus is to get handshake/application data decryption to work, 0-rtt data is for the future).

(Future work may also require the exporter secret, needed for DTLS-SRTP.)

 [1]: http://security.stackexchange.com/a/42350/2630

richsalz commented:
Thanks.  @levitte please re-confirm or, if you prefer, merge it yourself :)\
+1

levitte commented:
I'll merge

levitte commented:
Merged!
